lib/EnzymeCore/src/EnzymeCore.jl

```diff
@@ -189,50 +189,172 @@ end
     abstract type ABI
 
 Abstract type for what ABI  will be used.
+
+# Subtypes
+
+- [`FFIABI`](@ref) (the default)
+- [`InlineABI`](@ref)
+- [`NonGenABI`](@ref)
 """
 abstract type ABI end
 
 """
     struct FFIABI <: ABI
 
-Foreign function call ABI. JIT the differentiated function, then inttoptr call the address.
+Foreign function call [`ABI`](@ref). JIT the differentiated function, then inttoptr call the address.
 """
 struct FFIABI <: ABI end
+
 """
     struct InlineABI <: ABI
 
-Inlining function call ABI. 
+Inlining function call [`ABI`](@ref). 
 """
 struct InlineABI <: ABI end
+
 """
     struct NonGenABI <: ABI
 
-Non-generated function ABI. 
+Non-generated function [`ABI`](@ref). 
 """
 struct NonGenABI <: ABI end
+
 const DefaultABI = FFIABI
 
 """
-    abstract type Mode
+    abstract type Mode{ABI,ErrIfFuncWritten,RuntimeActivity}
+
+Abstract type for which differentiation mode will be used.
 
-Abstract type for what differentiation mode will be used.
+# Subtypes
+
+- [`ForwardMode`](@ref)
+- [`ReverseMode`](@ref)
+- [`ReverseModeSplit`](@ref)
+
+# Type parameters
+
+- `ABI`: what runtime [`ABI`](@ref) to use
+- `ErrIfFuncWritten`: whether to error when the function differentiated is a closure and written to.
+- `RuntimeActivity`: whether to enable runtime activity (default off)
+
+!!! warning
+    The type parameters of `Mode` are not part of the public API and can change without notice.
+    You can modify them with the following helper functions:
+    - [`WithPrimal`](@ref) / [`NoPrimal`](@ref)
+    - [`set_err_if_func_written`](@ref) / [`clear_err_if_func_written`](@ref)
+    - [`set_runtime_activity`](@ref) / [`clear_runtime_activity`](@ref)
+    - [`set_abi`](@ref)
 """
 abstract type Mode{ABI, ErrIfFuncWritten, RuntimeActivity} end
 
 """
-    struct ReverseMode{ReturnPrimal,RuntimeActivity,ABI,Holomorphic,ErrIfFuncWritten} <: Mode{ABI, ErrIfFuncWritten, RuntimeActivity}
+    set_err_if_func_written(::Mode)
+
+Return a new mode which throws an error for any attempt to write into an unannotated function object.
+"""
+function set_err_if_func_written end
+
+"""
+    clear_err_if_func_written(::Mode)
+
+Return a new mode which doesn't throw an error for attempts to write into an unannotated function object.
+"""
+function clear_err_if_func_written end
+
+"""
+    set_runtime_activity(::Mode)
```

Comment:
This also has a version which takes a bool [whether to turn on or off] as the second param, as well as for a FwdConfig and RevConfig, if those can be doc'd as well

Reply:
done